chirp/drivers/kgq10h.py

```diff
@@ -0,0 +1,3203 @@
+            rs_end = RadioSettingValueInteger(1, 999, s_grp.scan_end)
+            rs = RadioSetting("scn_grps.addrs[%i].scan_end" % i,
+                              "     Group %i End Channel" % (i + 1), rs_end)
+            rs.set_apply_callback(apply_end, s_grp)
+            scan.append(rs)
+        return scan
+
+    def _set_tone(self, mem, _mem):
+        def _set_dcs(code, pol):
+            val = int("%i" % code, 8) | 0x4000
+            if pol == "R":
+                val += 0x2000
+            return val
+
+        rx_mode = tx_mode = None
+        rxtone = txtone = 0x0000
+
+        if mem.tmode == "Tone":
+            tx_mode = "Tone"
+            rx_mode = None
+            txtone = int(mem.rtone * 10) + 0x8000
+        elif mem.tmode == "TSQL":
+            rx_mode = tx_mode = "Tone"
+            rxtone = txtone = int(mem.ctone * 10) + 0x8000
+        elif mem.tmode == "DTCS":
+            tx_mode = rx_mode = "DTCS"
+            txtone = _set_dcs(mem.dtcs, mem.dtcs_polarity[0])
+            rxtone = _set_dcs(mem.dtcs, mem.dtcs_polarity[1])
+        elif mem.tmode == "Cross":
+            tx_mode, rx_mode = mem.cross_mode.split("->")
+            if tx_mode == "DTCS":
+                txtone = _set_dcs(mem.dtcs, mem.dtcs_polarity[0])
+            elif tx_mode == "Tone":
+                txtone = int(mem.rtone * 10) + 0x8000
+            if rx_mode == "DTCS":
+                rxtone = _set_dcs(mem.rx_dtcs, mem.dtcs_polarity[1])
+            elif rx_mode == "Tone":
+                rxtone = int(mem.ctone * 10) + 0x8000
+
+        _mem.rxtone = rxtone
+        _mem.txtone = txtone
+
+        LOG.debug("Set TX %s (%i) RX %s (%i)" %
+                  (tx_mode, _mem.txtone, rx_mode, _mem.rxtone))
+
+    def set_memory(self, mem):
+        number = mem.number
+
+        _mem = self._memobj.memory[number]
+        _nam = self._memobj.names[number]
+
+        if mem.empty:
+            _mem.set_raw("\x00" * (_mem.size() // 8))
+            self._memobj.valid[number] = 0
+            self._memobj.names[number].set_raw("\x00" * (_nam.size() // 8))
+            return
+
+        _mem.rxfreq = int(mem.freq / 10)
+        if mem.duplex == "off":
+            # _mem.txfreq = 0xFFFFFFFF
+            _mem.txfreq = 0x00000000
+        elif mem.duplex == "split":
+            _mem.txfreq = int(mem.offset / 10)
+        elif mem.duplex == "off":
+            for i in range(0, 4):
+                _mem.txfreq[i].set_raw("\xFF")
+        elif mem.duplex == "+":
+            _mem.txfreq = int(mem.freq / 10) + int(mem.offset / 10)
+        elif mem.duplex == "-":
+            _mem.txfreq = int(mem.freq / 10) - int(mem.offset / 10)
+        else:
+            _mem.txfreq = int(mem.freq / 10)
+        _mem.scan_add = int(mem.skip != "S")
+
+        if ((mem.mode == "FM") or (mem.mode == "NFM")):
+            _mem.iswide = int((mem.mode == "FM"))
+            _mem.am_mode = 0
+        elif mem.mode == "AM":
+            _mem.am_mode = 1
```

Comment:
If I edit the memory in the UI and change mode to AM, the UI sends a memory like this to the driver:
```
mem.mode = "AM"
mem.extra["am_mode"] = "OFF"
```
So then on this L1857, you set `am_mode=1`. Then see below.

Reply:
So the UI is forcing am_mode to "off"?   I think that is my problem.   am_mode is the only setting available in the radio to turn AM on.   So it seems like the UI is workng against me here.

Reply:
The UI is not forcing anything. In spreadsheet view only one thing can change at a time. Meaning the user can only change mode, or am_mode, in a given edit and you will receive a `set_memory()` each time something changes. Your driver is the only thing that can (and should) enforce that things are consistent in a way that the radio requires. The UI has no idea that those two "columns" are related or dependent. In this way, making sure it works both in the spreadsheet and the properties view helps to ensure that you're doing the right thing in your driver, and not requiring the human user to *know* which combinations work and don't work when they're using the properties dialog.